README.md
# `odcr`

<!-- badges: start -->
[![CRAN version](https://www.r-pkg.org/badges/version/odcr)](https://CRAN.R-project.org/package=odcr)
[![R-CMD-check](https://github.com/eo2cube/odcr/workflows/R-CMD-check/badge.svg)](https://github.com/eo2cube/odcr/actions)
[![Package dependencies](https://tinyverse.netlify.com/badge/odcr)](https://CRAN.R-project.org/package=odcr)
<!-- badges: end -->

`odcr` is an R package that serves as a simple interface to the *Open Data Cube*. As such, it facilitates interaction with an *Open Data Cube* instance to access, query, list and load data as a native `xarray`-equivalent class. It implements basic methods to subset/index, plot and execute basic arithmetic operations on data and convert to native spatial raster classes such as `stars` or `raster`.

## Features (in development)

- [x] basic database connection (see `?odcr::database`)
- [x] `datacube` core `Datacube` class methods (see `?odcr::datacube`)
- [x] `datacube` core data discovery methods (see `?odcr::datacube`
- [x] `datacube` core data loading methods (see `?odcr::datacube`
- [ ] `datacube` core grid processing methods
- [ ] `datacube` core indexing methods
- [ ] support for native proxy class for streaming values directly (such as `stars_proxy`)
- [x] `xarray` `xarray.core.dataarray.DataArray` class and formal S3 definition
- [x] `xarray` `xarray.core.dataset.Dataset` class and formal S3 definition
- [x] `xarray` native S3 methods for retrieving dimensions (see `?odcr::dim`)
- [x] `xarray` native S3 methods for indexing by arbitrary number of dimensions (see `?odcr::Extract`)
- [x] `xarray` native S3 methods for indexing by variable (see `?odcr::Extract`)
- [x] `xarray` native S3 methods for indexing by time strings (see `?odcr::Extract`)
- [x] `xarray` native S3 methods for assigning/replacing variables (see `?odcr::Extract`)
- [x] `xarray` native S3 methods for arithmetic operators (see `?odcr::Arithmetic`)
- [x] `xarray` native S3 methods for coercing to native spatial classes `raster*` and `stars` (see `?odcr::as`)
- [x] `xarray` native S3 methods for plotting (see `?odcr::plot`)
- [ ] unit tests
- [x] configuration (see `?odcr::config`)
- [ ] ...


## System requirements

`odcr` requires the *Open Data Cube* Python library `datacube` which needs to be installed as part of a working *Open Data Cube* installation. See [odcbox](https://github.com/eo2cube/odcbox), if you want to install and set-up an *Open Data Cube* container environment that meets all `odcr` system requirements "out-of-the-box".

In case you have multiple Python installations or (virtual) environments installed, use `odcr::config` to make sure that `odcr` is using the correct Python installation and environment so that it can load and interact with the `datacube` library.

## Installation

`odcr` is in active development and thus breaking changes to its interface need to be expected. Install the latest development version of `odcr` from GitHub:

```r
devtools::install_github("eo2cube/odcr")
```

## Build instructions

In case you want to build the package manually, clone the repository and `cd` to its main directory. `odcr` uses `roxygen2` for documentation. Make sure, `devtools`, `roxygen2` and `pkgdown` are installed. Run:

```R
source("build.R")
```

## Get Started

See the vignette [*Using odcr*](https://eo2cube.github.io/odcr/articles/using_odcr.html) for a full introduction on how to use `odcr` with Sentinel-2 example data. Documentation and examples are available at [https://eo2cube.github.io/odcr/](https://eo2cube.github.io/odcr/).

## List of functions and methods

See [the package index](https://eo2cube.github.io/odcr/reference/index.html).






NEWS.md
# `odcr` changelog

## 0.0.2 (in development)

This release implements additional `xarray` features.

**Features:**

- `xarray` native S3 methods for assigning/replacing variables (see `?odcr::Extract`)

**Bug fixes:**

- The `xarray` attribute "unit" and the `xarray` name of the first object used in an arithmetic operation are not copied to the resulting `xarray` object anymore but stripped instead.


***

## 0.0.1 (pre-release)

This release implements a basic API for interacting with `datacube` and `xarray`.

**Features:**

- basic database connection (see `?odcr::database`)
- `datacube` core `Datacube` class methods (see `?odcr::datacube`)
- `datacube` core data discovery methods (see `?odcr::datacube`
- `datacube` core data loading methods (see `?odcr::datacube`
- `xarray` `xarray.core.dataarray.DataArray` class and formal S3 definition
- `xarray` `xarray.core.dataset.Dataset` class and formal S3 definition
- `xarray` native S3 methods for retrieving dimensions (see `?odcr::dim`)
- `xarray` native S3 methods for indexing by arbitrary number of dimensions (see `?odcr::Extract`)
- `xarray` native S3 methods for indexing by variable (see `?odcr::Extract`)
- `xarray` native S3 methods for indexing by time strings (see `?odcr::Extract`)
- `xarray` native S3 methods for arithmetic operators (see `?odcr::Arithmetic`)
- `xarray` native S3 methods for coercing to native spatial classes `raster*` and `stars` (see `?odcr::as`)
- `xarray` native S3 methods for plotting (see `?odcr::plot`)
- configuration (see `?odcr::config`)





R/config.R
#' Configure `odcr`
#'
#' This function configures `odcr` to connect with the correct Python environment and check the availability of the `datacube` Python library.
#'
#' @param python character, path to a python or conda binary/environment that should be used by `odcr`. Note that the selected Python environment must have the `datacube` library installed (and its upstream Python and system dependencies must be available). To set up a datacube environment ready for Pyhton and R 'out of the box', please see [eo2cube's odcbox](https://github.com/eo2cube/odcbox).
#' @param required logical, whether the use of this python install should be forced even if major libraries are missing or not
#' @md
#'
#' @return
#' None
#'
#' @importFrom reticulate use_python py_config import
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' config(python = "/env/bin/python3")
#' }
#' @export
#' @name config
config <- function(python = NA, required = T){
  if(!is.na(python)){
    #Sys.setenv(RETICULATE_PYTHON = python)
    use_python(python, required)
  }
  out("'odcr' is using the following python configuration:\n")
  print(py_config())
  cat("\n")

  if(.dc_available()){
    out(paste0("Linking to datacube ", .dc_version(), ", 'odcr' configuration successfull."))
  } else{
    out("Python library 'datacube' cannot be loaded. Did you specify the correct Python binary/environment? Is the 'datacube' module installed for this Python binary/environment and working? Are its upstream dependencies installed? To set up a datacube environment ready for Pyhton and R 'out of the box', please see https://github.com/eo2cube/odcbox", type = 3)
  }
}


R/methods.R
#' select var with this method
#' @importFrom reticulate py_get_item
#' @keywords internal
#' @noRd
.index_var <- function(ds, ...) {
  i <- list(...)[[1]]

  if(is.numeric(i)){
    i <- .get_measurements(ds)[i]
  }
  return(py_get_item(ds, i))
}

#' select dim with this method
#' @keywords internal
#' @noRd
.index_dim <- function(ds, i = NULL, j = NULL, k = NULL, ...) {
  index <- c(list(i, j, k), list(...))

  dims <- dim(ds)
  if(length(index) != length(dims)) out("Number of dimensions of object differ from provided index.", type = 3)
  names(index) <- names(dims)
  index <- index[!sapply(index, is.null)]

  if(any(!sapply(index, is.numeric))) out("Index must be numeric for this method.", type = 3)
  index <- lapply(index, function(i) as.integer(i-1))

  # check out of bounds! currently on the Python side

  do.call(ds$isel, index)
}

#' get dimensions
#' @keywords internal
#' @noRd
.dim <- function(ds) {
  x <- gsub("'", "", strsplit(strsplit(as.character(ds$sizes), "\\{")[[1]][2], ",")[[1]])
  x <- sapply(x, function(.x){
    gsub("\\)", "", gsub("\\}", "", .x))
  }, USE.NAMES = F)

  dims <- unlist(lapply(x, function(.x) as.numeric(strsplit(.x, ": ")[[1]][2])))
  names(dims) <- unlist(lapply(x, function(.x) trimws(strsplit(.x, ": ")[[1]][1])))
  dims <- dims[order(names(dims))]
  return(dims)
}

#' @noRd
#' @export
"[[.xarray.core.dataset.Dataset" <- .index_var

#' @noRd
#' @export
"[[.xarray.core.dataarray.DataArray" <- function(ds) return(ds)

#' @noRd
#' @export
"[.xarray.core.dataset.Dataset" <- .index_dim

#' @noRd
#' @export
"[.xarray.core.dataarray.DataArray" <- .index_dim

#' dim method
#' @noRd
#' @export
dim.xarray.core.dataset.Dataset <- .dim

#' dim method
#' @noRd
#' @export
dim.xarray.core.dataarray.DataArray <- .dim

#' plot method
#' @noRd
#' @export
plot.xarray.core.dataset.Dataset <- function(ds, ...) {
  plot(.xarray_convert(ds))
}

#' plot method
#' @noRd
#' @export
plot.xarray.core.dataarray.DataArray <- function(ds, ...) {
  arg <- list(...)
  if(is.null(arg$main)) arg$main <- ds$name
  do.call(plot, c(list(x = .xarray_convert(ds)), arg))
}

#' add method
#' @noRd
#' @export
"+.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  np$add(xds, yds)
}

#' substract method
#' @noRd
#' @export
"-.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  np$subtract(xds, yds)
}

#' divide method
#' @noRd
#' @export
"/.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  np$divide(xds, yds)
}

#' divide method
#' @noRd
#' @export
"*.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  np$multiply(xds, yds)
}

#' "xarray.core.dataset.Dataset" class
#'
#' @name xarray.core.dataset.Dataset-class
#' @aliases xarray.core.dataset.Dataset
#' @family xarray.core.dataset.Dataset
#'
#' @exportClass xarray.core.dataset.Dataset
setOldClass("xarray.core.dataset.Dataset")

#' Methods to coerce \code{odcr} classes to native spatial classes
#' @name as
#' @rdname coerce-methods
#' @aliases coerce,xarray.core.dataset.Dataset,stars-method.
#' @exportMethod coerce
setAs("xarray.core.dataset.Dataset", "stars", function(from) as.stars(from))


#' "xarray.core.dataarray.DataArray" class
#'
#' @name xarray.core.dataarray.DataArray-class
#' @aliases xarray.core.dataarray.DataArray
#' @family xarray.core.dataarray.DataArray
#'
#' @exportClass xarray.core.dataarray.DataArray
setOldClass("xarray.core.dataarray.DataArray")

#' Methods to coerce \code{odcr} classes to native spatial classes
#' @name as
#' @rdname coerce-methods
#' @aliases coerce,xarray.core.dataarray.DataArray,stars-method.
#' @exportMethod coerce
setAs("xarray.core.dataarray.DataArray", "stars", function(from) as.stars(from))

#' Methods to coerce \code{odcr} classes to native spatial classes
#'
#' [as.stars()] allows to convert `xarray.core.dataset.Dataset` to `stars`.
#'
#' @rdname coerce-methods
#' @name as.stars
#' @md
#'
#' @param from object of class `xarray.core.dataset.Dataset`
#'
#' @return [as.stars()] retruns an object of class `stars`
#' @export
as.stars <- function(from){
  .xarray_convert(from, method = "stars")
}

#' Methods to coerce \code{odcr} classes to native spatial classes
#'
#' [as.raster()] allows to convert `xarray.core.dataset.Dataset` to `RasterLayer` or `RasterStack`.
#'
#' @rdname coerce-methods
#' @name as.raster
#' @md
#'
#' @param from object of class `xarray.core.dataset.Dataset`
#'
#' @return [as.raster()] retruns an object of class `RasterLayer` or `RasterStack`
#' @export
as.raster <- function(from){
  .xarray_convert(from, method = "raster")
}



R/database.R
#' Connect to a database
#'
#' These functions connect to a database that houses a datacube.
#'
#' @param ... optional arguments to index or configure a connection.
#' @param object logical, whether the connection should be returned or not.
#'
#' @return
#' None or the connection of class 'datacube.api.core.Datacube', if \code{object = T}
#'
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' # connect to a database, store the Daatcube connection internally (default and recommended)
#' database_connect(app = "Sentinel_2")
#'
#' # retrieve the current connection as an object
#' dc <- database_get_active()
#'
#' # or return a Datacube object
#' dc <- database_connect(app = "Sentinel_2", object = TRUE)
#'
#'
#' }
#' @export
#' @name database
database_connect <- function(..., object = F){
  if(is.environment(getOption("odcr.dc"))) out("Exisiting datacube connection is being replaced.", type = 2)
  options("odcr.dc" = datacube$Datacube(...))

  if(isTRUE(object)) return(getOption("odcr.dc"))
}

#' @rdname database
#' @export
database_get_active <- function(){
  return(getOption("odcr.dc"))
}


R/datacube.R
#' Query a datacube
#'
#' These functions query a datacube to which a connection has been established using \code{\link{database_connect}}.
#'
#' @param dc a datacube connection of class 'datacube.api.core.Datacube'. By default, the session connection initialized using \code{connect()} is used.
#' @param query list, containing query parameters such as
#' \itemize{
#'    \item \code{'product'}, character, name of the product to be queries (see \code{dc_list_products()})
#'    \item \code{'time'}, character vector of length == 2, containing a time query
#'    \item \code{'x'}, numeric vector of length == 2, containing xmin and xmax
#'    \item \code{'y'}, numeric vector of length == 2, containing ymin and ymax
#'    \item \code{'output_crs'}, character, the output CRS, e.g. as epsg in the format \code{'epsg:6933'}
#'    \item \code{'resolution'}, numeric vector of length == 2
#'    \item \code{'measurements'}, character vector containing the names or aliases of measurements
#' }
#' @param lazy logical, whether to use lazy dataset finding or not.
#'
#' @return
#' None or \code{data.frame}
#'
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' # connect to a database
#' database_connect(app = "Sentinel_2")
#'
#' # return a data.frame containing all stored variables/measurements of the current connection
#' # and its aliases, data types etc.
#' dc_list_measurements()
#'
#' # return a data.frame containing all products of the current connection
#' dc_list_products()
#'
#' # build a query list
#' lat <- 22.821
#' lon <- 28.518
#' buffer <- 0.05
#'
#' query <- list(
#'   'time' = c('2020-01', '2020-03'),
#'   'x' = c(lon - buffer, lon + buffer),
#'   'y' = c(lat + buffer, lat - buffer),
#'   'output_crs' = 'epsg:6933',
#'   'resolution' = c(-20,20)
#' )
#'
#' # return a data.frame of all datasets matching the query
#' dc_find_datasets(query = c(product = "s2_l2a", query))
#'
#' # load data and return an xarray object for a query
#' ds <- dc_load(query = c(product = "s2_l2a", dask_chunks = dict(), query))
#'
#' # after a lot of queries, you may want to close open connections
#' dc_close()
#' }
#' @export
#' @name datacube
dc_close <- function(dc = getOption("odcr.dc")){
  if(!is.environment(dc)) out("Connect to a datacube to close open connections.", type = 3)
  .check_class(dc, "dc", "datacube.api.core.Datacube")

  dc$close()
}

#' @rdname datacube
#' @export
dc_find_datasets <- function(dc = getOption("odcr.dc"), query, lazy = FALSE){
  if(!is.environment(dc)) out("Connect to a datacube to find datasets.", type = 3)
  .check_class(dc, "dc", "datacube.api.core.Datacube")

  # transform query
  query <- .transform_query(dc, query)

  # query datasets
  if(isFALSE(lazy)){
    ds_list <- do.call(dc$find_datasets, query)
  } else{
    ds_list <- do.call(dc$find_datasets_lazy, query)
  }

  # transpose paths
  ds_paths <- as.data.frame(t(sapply(ds_list, function(x){
    paths <- sapply(x$measurements, function(y) y$path)

    if(!is.null(query$measurements)){
      return(paths[query$measurements])
    } else{
      return(paths)
    }
  })))

  return(ds_paths)
}

#' @rdname datacube
#' @export
dc_list_measurements <- function(dc = getOption("odcr.dc")){
  if(!is.environment(dc)) out("Connect to a datacube to list measurements.", type = 3)
  .check_class(dc, "dc", "datacube.api.core.Datacube")
  as.data.frame(dc$list_measurements(), stringAsFactors = F)
}

#' @rdname datacube
#' @export
dc_list_products <- function(dc = getOption("odcr.dc")){
  if(!is.environment(dc)) out("Connect to a datacube to list products.", type = 3)
  .check_class(dc, "dc", "datacube.api.core.Datacube")
  as.data.frame(dc$list_products(), stringAsFactors = F)
}

#' @rdname datacube
#' @importFrom reticulate dict
#' @export
dc_load <- function(dc = getOption("odcr.dc"), query){
  .check_class(dc, "dc", "datacube.api.core.Datacube")

  if(is.null(query$dask_chunks)) query$dask_chunks <- dict()
  ds <- do.call(dc$load, query)
  return(ds)
}



R/xarray_methods.R
#' convert index into name
#' @keywords internal
#' @noRd
.index2name <- function(x, ...){
  i <- list(...)[[1]]
  if(is.numeric(i)){
    i <- .get_measurements(x)[i]
  }
  return(i)
}

#' select var with this method
#' @importFrom reticulate py_get_item
#' @keywords internal
#' @noRd
.index_var <- function(x, ...) {
  i <- .index2name(x, ...)
  return(py_get_item(x, i))
}

#' assign/set var with this method
#' @importFrom reticulate py_set_item
#' @keywords internal
#' @noRd
.set_var <- function(x, ..., value){
  i <- .index2name(x, ...)
  py_set_item(x, i, value)
  return(x)
}

#' select dim with this method
#' @keywords internal
#' @noRd
.index_dim <- function(x, i = NULL, j = NULL, k = NULL, ...) {
  index <- c(list(i, j, k), list(...))

  dims <- dim(x)
  if(length(index) != length(dims)) out("Number of dimensions of object differ from provided index.", type = 3)
  names(index) <- names(dims)
  index <- index[!sapply(index, is.null)]

  if(any(!sapply(index, is.numeric))) out("Index must be numeric for this method.", type = 3)
  index <- lapply(index, function(i) as.integer(i-1))

  # check out of bounds! currently on the Python side

  do.call(x$isel, index)
}

#' get dimensions
#' @keywords internal
#' @noRd
.dim <- function(x) {
  x <- gsub("'", "", strsplit(strsplit(as.character(x$sizes), "\\{")[[1]][2], ",")[[1]])
  x <- sapply(x, function(.x){
    gsub("\\)", "", gsub("\\}", "", .x))
  }, USE.NAMES = F)

  dims <- unlist(lapply(x, function(.x) as.numeric(strsplit(.x, ": ")[[1]][2])))
  names(dims) <- unlist(lapply(x, function(.x) trimws(strsplit(.x, ": ")[[1]][1])))
  dims <- dims[order(names(dims))]
  return(dims)
}

#' @title Methods to extract from or assign to \code{odcr} classes
#'
#' @description `[` allows to subset an `xarray` object by its dimensions (see [`dim()`])
#'
#' @name Extract
#' @md
#'
#' @param i numeric, first dimension index
#' @param j numeric, second dimension index
#' @param k numeric, third dimension index
#'
#' @return Subsetted `xarray` object
#'
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' # connect to a database, store the Daatcube connection internally (default and recommended)
#' database_connect(app = "Sentinel_2")
#'
#' # build a query list
#' lat <- 22.821
#' lon <- 28.518
#' buffer <- 0.05
#'
#' query <- list(
#'   'time' = c('2020-01', '2020-03'),
#'   'x' = c(lon - buffer, lon + buffer),
#'   'y' = c(lat + buffer, lat - buffer),
#'   'output_crs' = 'epsg:6933',
#'   'resolution' = c(-20,20)
#' )
#'
#' # load data and return an xarray object for a query
#' ds <- dc_load(query = c(product = "s2_l2a", dask_chunks = dict(), query))
#'
#' # return dimensions of ds
#' dim(ds)
#'
#' # subset/index ds by time, x, y
#' ds[13, 11:20, 21:30]
#'
#' # subset/index ds by time
#' ds[13,,]
#'
#' # subset/index ds by variable/measurement
#' ds[[2]]
#'
#' # or do the same with a named variable or alias
#' ds[["nir"]]
#'
#' # subset/index by exact time vectors
#' xar_sel_time(ds, c("2020-01-02", "2020-01-07"))
#'
#' # or find closest times/date subsets by time vectors
#' xar_sel_time(ds, c("2020-01-01", "2020-01-07"), exact_match = F)
#' }
#' @export
"[.xarray.core.dataset.Dataset" <- .index_dim

#' @rdname Extract
#' @export
"[.xarray.core.dataarray.DataArray" <- .index_dim


#' @description `[[` allows to subset an `xarray` object by measurements/variables (e.g. spectral band), either named (character) or indexed (numeric).
#'
#' @rdname Extract
#' @md
#'
#' @param ... numeric or character, index or indices by which to extract/assign (additional) elements
#'
#' @export
"[[.xarray.core.dataset.Dataset" <- .index_var

#' @rdname Extract
#' @export
"[[.xarray.core.dataarray.DataArray" <- function(x, ...) return(x)

#' @description `xar_sel_time()` allows to subset an `xarray` object by a character vector of dates/times.
#' @rdname Extract
#' @md
#'
#' @param x `xarray` object, the dataset to be subsetted from or assigned to
#' @param query character vector, one or more time/date character vectors in the format of the time dimension of `x` or an abbreviated form of it (e.g. only the date component)
#' @param exact_match logical, whether to return only exact matches (default) or to search for closest elements to each element in `query` using `[difftime()]`
#'
#' @export
xar_sel_time <- function(x, query, exact_match = T){
  v <- x$time$values
  if(isTRUE(exact_match)){
    subs <- sapply(query, function(.q) which(grepl(.q, v)), USE.NAMES = F)
    if(length(subs) == 0) out("Query times are not matching times of x.", type = 3)
    subs_l <- sapply(subs, length)
    if(all(subs_l == 0)) out("Query times are not matching times of x.", type = 3)
    if(any(subs_l == 0)) out("Some query times are not matching times of x, subsetting only by valid times.", type = 2)
    subs <- unlist(subs)
    x$sel(time = v[subs])
  }else{
    subs <- sapply(as.POSIXct(query), function(.q) which.min(abs(difftime(.q, as.POSIXct(v)))))
    if(any(duplicated(subs))) out("Output contains duplicated times, since some query times are closest to the same times of x.", type = 2)
    x$sel(time = v[subs])
  }
}

#' @description `[[<-` allows to assign a measurement/variable (e.g. spectral band), either named (character) or indexed (numeric), to an existing `xarray` object.
#' @md
#'
#' @param value `xarray` object, the dataset that should be assigned to `x`
#' @rdname Extract
#'
#' @export
"[[<-.xarray.core.dataset.Dataset" <- .set_var

#' @rdname Extract
#'
#' @export
"$<-.xarray.core.dataset.Dataset" <- .set_var


#' Dimensions of \code{odcr} classes
#'
#' `dim()` retrieves dimensions from an `xarray` object.
#'
#' @name dim
#' @md
#'
#' @param x object of which dimensions should be retrieved.
#'
#' @return a (named) vector
#'
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' # connect to a database, store the Daatcube connection internally (default and recommended)
#' database_connect(app = "Sentinel_2")
#'
#' # build a query list
#' lat <- 22.821
#' lon <- 28.518
#' buffer <- 0.05
#'
#' query <- list(
#'   'time' = c('2020-01', '2020-03'),
#'   'x' = c(lon - buffer, lon + buffer),
#'   'y' = c(lat + buffer, lat - buffer),
#'   'output_crs' = 'epsg:6933',
#'   'resolution' = c(-20,20)
#' )
#'
#' # load data and return an xarray object for a query
#' ds <- dc_load(query = c(product = "s2_l2a", dask_chunks = dict(), query))
#'
#' # return dimensions of ds
#' dim(ds)
#' }
#' @export
dim.xarray.core.dataset.Dataset <- .dim

#' dim method
#' @rdname dim
#' @export
dim.xarray.core.dataarray.DataArray <- .dim


#' Plotting of \code{odcr} classes
#'
#' `plot()` plots an `xarray` object using `[stars::plot()]`
#'
#' @name plot
#' @md
#'
#' @param x object which should be plotted
#' @param ... additional arguments passed to [`stars::plot()`]
#'
#' @return a `stars` plot
#'
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' # connect to a database, store the Daatcube connection internally (default and recommended)
#' database_connect(app = "Sentinel_2")
#'
#' # build a query list
#' lat <- 22.821
#' lon <- 28.518
#' buffer <- 0.05
#'
#' query <- list(
#'   'time' = c('2020-01', '2020-03'),
#'   'x' = c(lon - buffer, lon + buffer),
#'   'y' = c(lat + buffer, lat - buffer),
#'   'output_crs' = 'epsg:6933',
#'   'resolution' = c(-20,20)
#' )
#'
#' # load data and return an xarray object for a query
#' ds <- dc_load(query = c(product = "s2_l2a", dask_chunks = dict(), query))
#'
#' # plot first timestamp, spatial subset and only first band/variable
#' plot(ds[1,101:300,101:300][[1]])
#'
#' # plot first timestamp, spatial subset and only two bands/variables
#' plot(ds[1,101:300,101:300][[c(1,2)]])
#' }
#' @export
plot.xarray.core.dataset.Dataset <- function(x, ...) {
  plot(.xarray_convert(x), ...)
}

#' plot method
#' @rdname plot
#' @export
plot.xarray.core.dataarray.DataArray <- function(x, ...) {
  arg <- list(...)
  if(is.null(arg$main)) arg$main <- x$name
  do.call(plot, c(list(x = .xarray_convert(x)), arg))
}


#' @title Arithmetic operators for \code{odcr} classes
#'
#' @description These arithmetic operators perform arithmetic on two objects of the same class, shape and dimensions.
#'
#' @name Arithmetic
#' @md
#'
#' @param xds  `xarray` object
#' @param yds `xarray` object
#'
#' @return An `xarray` object of same class, shape and dimensions as the inputs to the operator, holding the result.
#'
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' # connect to a database, store the Daatcube connection internally (default and recommended)
#' database_connect(app = "Sentinel_2")
#'
#' # build a query list
#' lat <- 22.821
#' lon <- 28.518
#' buffer <- 0.05
#'
#' query <- list(
#'   'time' = c('2020-01', '2020-03'),
#'   'x' = c(lon - buffer, lon + buffer),
#'   'y' = c(lat + buffer, lat - buffer),
#'   'output_crs' = 'epsg:6933',
#'   'resolution' = c(-20,20)
#' )
#'
#' # load data and return an xarray object for a query
#' ds <- dc_load(query = c(product = "s2_l2a", dask_chunks = dict(), query))
#'
#' ndvi <- (ds[["nir"]] - ds[["red"]]) / (ds[["nir"]] + ds[["red"]])
#' ndvi
#'
#' }
#' @export
"+.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  y <- np$add(xds, yds)
  .rm_attr(.rm_name(y), "units")
}

#' @rdname Arithmetic
#' @export
"-.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  y <- np$subtract(xds, yds)
  .rm_attr(.rm_name(y), "units")
}

#' @rdname Arithmetic
#' @export
"/.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  y <- np$divide(xds, yds)
  .rm_attr(.rm_name(y), "units")
}

#' @rdname Arithmetic
#' @export
"*.xarray.core.dataarray.DataArray" <- function(xds, yds) {
  y <- np$multiply(xds, yds)
  .rm_attr(.rm_name(y), "units")
}

#' "xarray.core.dataset.Dataset" class
#'
#' @name xarray.core.dataset.Dataset-class
#' @aliases xarray.core.dataset.Dataset
#' @family xarray.core.dataset.Dataset
#'
#' @importFrom methods setOldClass
#' @exportClass xarray.core.dataset.Dataset
setOldClass("xarray.core.dataset.Dataset")

#' Methods to coerce \code{odcr} classes to native spatial classes
#' @name as
#' @rdname coerce-methods
#' @aliases coerce,xarray.core.dataset.Dataset,stars-method.
#'
#' @importFrom methods setAs
setAs("xarray.core.dataset.Dataset", "stars", function(from) as.stars(from))


#' "xarray.core.dataarray.DataArray" class
#'
#' @name xarray.core.dataarray.DataArray-class
#' @aliases xarray.core.dataarray.DataArray
#' @family xarray.core.dataarray.DataArray
#'
#' @importFrom methods setOldClass
#' @exportClass xarray.core.dataarray.DataArray
setOldClass("xarray.core.dataarray.DataArray")

#' Methods to coerce \code{odcr} classes to native spatial classes
#' @name as
#' @rdname coerce-methods
#' @aliases coerce,xarray.core.dataarray.DataArray,stars-method.
#'
#' @importFrom methods setAs
setAs("xarray.core.dataarray.DataArray", "stars", function(from) as.stars(from))

#' Methods to coerce \code{odcr} classes to native spatial classes
#'
#' [as.stars()] allows to convert `xarray.core.dataset.Dataset` to `stars`.
#'
#' @rdname coerce-methods
#' @name as.stars
#' @md
#'
#' @param from object of class `xarray.core.dataset.Dataset`
#'
#' @return [as.stars()] retruns an object of class `stars`
#'
#' @examples
#' \dontrun{
#' library(odcr)
#'
#' # connect to a database, store the Daatcube connection internally (default and recommended)
#' database_connect(app = "Sentinel_2")
#'
#' # build a query list
#' lat <- 22.821
#' lon <- 28.518
#' buffer <- 0.05
#'
#' query <- list(
#'   'time' = c('2020-01', '2020-03'),
#'   'x' = c(lon - buffer, lon + buffer),
#'   'y' = c(lat + buffer, lat - buffer),
#'   'output_crs' = 'epsg:6933',
#'   'resolution' = c(-20,20)
#' )
#'
#' # load data and return an xarray object for a query
#' ds <- dc_load(query = c(product = "s2_l2a", dask_chunks = dict(), query))
#' ds <- ds[,101:300,101:300]
#'
#' # convert to stars (multi-dimensional and mulit-varariable)
#' ds_stars <- as.stars(ds)
#' ds_stars <- as(ds, "stars")
#'
#' # convert to stars (single variable)
#' ds_stars <- as.stars(ds[[1]])
#'
#' # convert to stars (2-dim, multi-variable)
#' ds_stars <- as.stars(ds[1,,])
#'
#' # convert to raster (multi-dimensional and mulit-varariable)
#' ds_raster <- as.raster(ds)
#' ds_raster <- as(ds, "raster")
#' # raster cannot represent 4-dim objects, thus coereced to list of RasterBricks
#'
#' # convert to starasterrs (single variable)
#' ds_raster <- as.raster(ds[[1]])
#'
#' # convert to raster (2-dim, multi-variable)
#' ds_raster <- as.raster(ds[1,,])
#' }
#'
#' @export
as.stars <- function(from){
  .xarray_convert(from, method = "stars")
}

#' Methods to coerce \code{odcr} classes to native spatial classes
#'
#' [as.raster()] allows to convert `xarray.core.dataset.Dataset` to `RasterLayer` or `RasterBrick`.
#'
#' @rdname coerce-methods
#' @name as.raster
#' @md
#'
#' @param from object of class `xarray.core.dataset.Dataset`
#'
#' @return [as.raster()] retruns an object of class `RasterLayer` or `RasterBrick` or a `list` of such in case of more than 3 dimensions
#'
#' @importFrom utils installed.packages
#' @export
as.raster <- function(from){
  if(!any("raster" == installed.packages())) out("Package 'raster' must be installed to convert to 'Raster*'", type = 3)
  .xarray_convert(from, method = "raster")
}

